core-rust/state-manager/src/mempool/mempool_manager.rs

```diff
@@ -66,25 +66,36 @@ use super::metrics::MempoolManagerMetrics;
 use crate::mempool::priority_mempool::*;
 use crate::mempool::*;
 use crate::MempoolAddSource;
+use crate::StateVersion;
+use node_common::locks::Mutex;
 use node_common::metrics::TakesMetricLabels;
 use prometheus::Registry;
 use transaction::model::*;
 
-use std::cmp::max;
 use std::collections::HashSet;
+use std::sync::mpsc::Receiver;
 use std::sync::Arc;
 use std::time::Instant;
 
 use crate::mempool_relay_dispatcher::MempoolRelayDispatcher;
 use crate::store::StateManagerDatabase;
-use crate::transaction::{
-    CachedCommittabilityValidator, ForceRecalculation, PrevalidatedCheckMetadata,
-};
+use crate::transaction::{CachedCommittabilityValidator, ForceRecalculation};
 use node_common::locks::RwLock;
 use tracing::warn;
 
+pub enum MempoolUpdate {
+    StateVersionUpdate(StateVersionUpdate),
+    TransactionRejected(NotarizedTransactionHash),
+}
+
+pub struct StateVersionUpdate {
+    pub payload_hash: NotarizedTransactionHash,
+    pub state_version: StateVersion,
+}
+
 /// A high-level API giving a thread-safe access to the `PriorityMempool`.
 pub struct MempoolManager {
+    deferred_updates_rx: Mutex<Receiver<MempoolUpdate>>,
```

Comment:
_(note from author)_

The `Mutex<>` is needed solely to circumvent `jni`s `Sync` requirement. The overhead is just the extra memory access (+ compare) since it will never actually wait.
In a pure Rust environment this is not needed and should be removed in future.

Reply:
plz paste this note's contents into `// TODO(after 100% Rust migration): ...`

Reply:
anyways - is there a conceptual reason for "deferring" updates?
AFAI-vaguely-R from our Slack conversations, the Tx/Rx exists only to break some cyclic dep between the mempool and PTRCache...? If this is the case, then at the very least we need a fat comment about this in the code. Also, please give me a moment to play with it (and possibly invent some other hack for the cyclic dep, _once you confirm you don't need deferring for other reasons_)

Reply:
>AFAI-vaguely-R from our Slack conversations, the Tx/Rx exists only to break some cyclic dep between the mempool and PTRCache...? If this is the case, then at the very least we need a fat comment about this in the code. Also, please give me a moment to play with it (and possibly invent some other hack for the cyclic dep, once you confirm you don't need deferring for other reasons)

This is the main reason for the current goal. Either we fully tie result cache with mempool (which would add more contention) or we do this decoupling (otherwise it's impossible since we get circular dependencies). I am pretty sure we will end up having an actor based model where `MempoolManager` will have it's own thread, in which case this is no longer "deferred" work, but the channel actually used as a proper channel.

Reply:
There is also the option of redesigning the mempool and result cache. I think it makes more sense to have the results of transactions in mempool to live in the mempool and leave result cache for rejected transactions and committable transactions that don't make it into mempool (full mempool).

Reply:
@malexandru-rdx I think we can now un-pause this PR.

I gave it a try to model the dependencies here in some other way, and my suggestion is at https://github.com/radixdlt/babylon-node/pull/696 - please have a look there and feel free to merge it into this PR if you find it beneficial.